lib/Mojolicious/Plugin/DefaultHelpers.pm

```diff
@@ -161,6 +165,61 @@ sub _is_fresh {
   return $c->app->static->is_fresh($c, \%options);
 }
 
+sub _proxy_method_p {
+  my ($method, $c) = (shift, shift);
+  return _proxy_start_p($c, $c->ua->build_tx($method, @_));
+}
+
+sub _proxy_start_p {
+  my ($c, $source_tx, $cb) = @_;
+  my $tx = $c->render_later->tx;
+
+  my $promise = Mojo::Promise->new;
+  $source_tx->res->content->auto_upgrade(0)->auto_decompress(0)->once(
+    body => sub {
+      my $source_content = shift;
+
+      my $source_res = $source_tx->res;
+      my $res        = $tx->res;
+      my $content    = $res->content;
+      $res->code($source_res->code)->message($source_res->message);
+      my $headers = $source_res->headers->clone->dehop;
+      $content->headers($headers);
+      $c->$cb($source_tx) if $cb;
+      $promise->resolve();
+
+      my $source_stream = Mojo::IOLoop->stream($source_tx->connection);
+      return unless my $stream = Mojo::IOLoop->stream($tx->connection);
+
+      my $write = $source_content->is_chunked ? 'write_chunk' : 'write';
+      $source_content->unsubscribe('read')->on(
+        read => sub {
+          $content->$write(pop) and $tx->resume;
+
+          # Throttle transparently when backpressure rises
+          return if $stream->can_write;
+          $source_stream->stop;
+          $stream->once(drain => sub { $source_stream->start });
```

Comment:
I think this throttling code is really nice 👍

Reply:
I like it too, but i wish `once` handlers were less expensive.